enhancements/20231113-witness-node-support.md

```diff
@@ -0,0 +1,93 @@
+
+## Summary
+
+The witness node is a lightweight node that only runs the control plane components. The witness node is not schedulable and does not run any workload. The witness node forms a quorum with the other two nodes.
+
+### Related Issues
+
+https://github.com/harvester/harvester/issues/3266
+
+## Motivation
+
+Some users do not have three powerful nodes for the edge uses case. Instead, they would have two powerful nodes and one lightweight node. The lightweight node is only used for the `etcd` role. Kubernetes need at least three nodes to form a quorum. The witness node support will help the edge uses case. That will benefit both cost and high availability.
+
+### Goals
+
+Create a three-node cluster with one witness node.
+
+## Proposal
+
+### User Stories
+
+For the edge use case, the witness node support benefits both cost and high availability. The classical scenario would be two nodes and one witness node.
+
+### User Experience In Detail
+
+1. Set the node label `node-role.harvesterhci.io/etcd=true` to the node.
+2. The promoter will add the node to the cluster as `etcd` role.
+
+### API changes
+
+None
+
+## Design
+
+### Limitations
+
+- Each cluster only supports one witness node.
+
+### Implementation Overview
+
+#### Promote controller
+
+Harvester depends on the job `promote` to promote a node to related role after it joins (the default role is worker). We can only promote the node to `etcd` role with the label `node-role.harvesterhci.io/etcd=true`. Also, we need to add taint for etcd role to prevent workload from scheduling to the node.
+
+
+After this feature, the harvester node will have the following roles:
+1. master node roles:
+   - node-role.kubernetes.io/control-plane: "true"
+   - node-role.kubernetes.io/etcd: "true"
+   - node-role.kubernetes.io/master: "true"
+
+2. etcd node roles:
+   - node-role.kubernetes.io/etcd: "true"
+
+3. worker node roles:
+
+**OPTIONAL**: Promote controller can change the node role dynamically.
+
+#### Requirements
+
+The hardware requirements for the witness node could be a reasonable amount. We might need to set the minimum requirements for the witness node.
+
+Because of that, we might need to update the documentation for both the minimum requirements of the witness node and master node.
+
+#### Installation
+
+1. Interactive mode:
+   - We need to add an extra panel for setting the witness node (only for join cluster mode).
+
+2. Non-interactive mode:
+   - We can simply the label `node-role.harvesterhci.io/etcd=true` to the node by yaml.
+
+#### Miscellaneous changes
+
+- **Longhorn**: 
+    
+   The default StorageClass `harvester-longhorn` will have three replicas, but it could not be satisfied with the witness node. We need to update the default StorageClass to two replicas.
+
+### Test plan
+
+1. Create a three-node cluster through both interactive and non-interactive modes.
+2. Check the node roles after the cluster is created. (Two master nodes and one etcd node)
```

Comment:
Question:
When there is a two-node cluster with 1 master + 1 witness,  e.g., the third-node is not installed yet, or gone, or removed, anything needs to be cared?

Reply:
Promote will be triggered on at least three nodes so we do not encounter this situation.